sql/src/main/java/io/crate/operation/collect/LuceneDocCollector.java

```diff
@@ -162,29 +152,58 @@ public void collect(int doc) throws IOException {
             e.setNextDocId(doc);
         }
         boolean wantMore = downstream.setNextRow(inputRow);
+
+        postCollectDoc(doc);
+
         if (!wantMore || (limit != null && rowCount == limit)) {
             // no more rows required, we can stop here
             throw new CollectionFinishedEarlyException();
         }
     }
 
+    protected boolean skipDoc(int doc) {
+        ScoreDoc after = internalCollectContext.lastCollected;
+        if (after != null) {
+            if (currentDocBase == internalCollectContext.lastDocBase && doc <= after.doc) {
+                // doc was already collected, skip
+                LOGGER.trace("skipping doc {} <= {} of docBase {}", doc, after.doc, currentDocBase);
+                return true;
+            }
+        }
+        return false;
+    }
+
+    protected void postCollectDoc(int doc) {
+        if (internalCollectContext.lastCollected == null) {
+            internalCollectContext.lastCollected = new ScoreDoc(doc, Float.NaN);
+        }
+        internalCollectContext.lastCollected.doc = doc;
+    }
+
+
     @Override
     public void setNextReader(AtomicReaderContext context) throws IOException {
-        this.currentReader = context.reader();
+        currentReader = context.reader();
+        currentDocBase = context.docBase;
+        skipSegmentReader();
         for (LuceneCollectorExpression expr : collectorExpressions) {
             expr.setNextReader(context);
         }
     }
 
-    public void setNextOrderByValues(ScoreDoc scoreDoc) {
-        for (OrderByCollectorExpression expr : orderByCollectorExpressions) {
-            expr.setNextFieldDoc((FieldDoc) scoreDoc);
+    protected void skipSegmentReader() {
+        if (currentDocBase < internalCollectContext.lastDocBase) {
+            // skip this segment reader, all docs of this segment are already collected
+            LOGGER.trace("skipping segment reader with docBase {} < {}",
+                    currentDocBase, internalCollectContext.lastDocBase);
+            throw new CollectionTerminatedException();
```

Comment:
This exception is catch at the IndexSearcher, it will move to next segment if thrown. So no this is different ;)